### ISSUE TITLE
Bootstrap 4 Tooltip Persists, Looks Terrible

### ISSUE THREAD

**Issue by dagoldman** [twbs/bootstrap#28964]
OS is Windows 7. Browsers showing problem include chrome 74 and IE 11 browsers.

Normally, bootstrap does not display a tooltip for a disabled button. However, when a button is disabled while the tooltip is being displayed, the tooltip persists. Of course this looks terrible. So when clicking a button results in the button being disabled, the tooltip does not go away.

This is either a bug, or there is some setting that I am not aware of that needs to be added to my source code. However, I have looked through the documentation and cannot find anything I am failing to add. Also, I have searched the internet, including this site, and cannot find an answer to this problem. However, it seems likely the problem must have previously happened to others.

I have included a complete test case below, with detailed comments. Could you please run the test case and at least see if the problem occurs when you run it?

Assuming the problem occurs, could you please verify if this is a bug, or I am just missing something? If the latter, what am I missing, and I would suggest adding to the docs, or making the default case automatically work. Any help appreciated. Thanks

```
<!DOCTYPE html>
<html lang="en">
<head>
<title>Bootstrap 4 Tooltip Persists, Looks Terrible</title>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
</head>
<body>
<div class="container">
<h3>Bootstrap 4 Tooltip Persists, Looks Terrible</h3>
<br> "Other button disabled" scenario works fine.
<br> First, hover over buttons. Look at tooltips while hovering.
<br> Then, click button A to disable button B. Works fine:
<br> - Button B is disabled, and no tooltip.
<br> - Tooltip still works correctly on button A.
<br> <br>
<button type="button" id="button_a" class="btn btn-primary" data-toggle="tooltip" data-trigger="hover" title="Disables Button B" onclick='change_button ("button_b", true);'>
Button A
</button>
<button type="button" id="button_b" class="btn btn-primary" data-toggle="tooltip" data-trigger="hover" title="Disabled by Button A">
Button B
</button>
<br> <br> <hr>
<br> "Same button disabled" scenario fails.
<br> First, hover over Button C. Look at the tooltip.
<br> Then, click button C to disable button C. Messes up:
<br> - Button C is disabled, so that part is OK.
<br> - But Button C tooltip persists, looks terrible.
<br> <br>
<button type="button" id="button_c" class="btn btn-primary" data-toggle="tooltip" data-trigger="hover" title="Disables Button C" onclick='change_button ("button_c", true);'>
Button C
</button>
<button type="button" id="button_d" class="btn btn-primary" data-toggle="tooltip" data-trigger="hover" title="Enables Button C" onclick='change_button ("button_c", false);'>
Button D
</button>
<br> <br> <hr> <br>
So if a button is disabled while the cursor is hovering over it, the tooltip persists.
<strong>
Could you please explain how to make tooltip go away when Button C is disabled?
</strong>
I would prefer if tooltip went away when Button C disabled,
since that is default behavior of bootstrap disabled buttons.
And disabled button may later get enabled again, so I do not want to delete title attribute.
Basically I want bootstrap to behave the same way as when Button B is disabled.
I have looked around and cannot find how to fix this problem,
and have tried various things, with no success.
There are real instances where a button carries out an action,
and then the button needs to be disabled, so I appreciate the help. Thanks
</div>
<script>
$(document).ready(function(){$('[data-toggle="tooltip"]').tooltip();});
function change_button (button_id, bool_val) {
var button_obj = document.getElementById (button_id);
button_obj.disabled = bool_val;
}
</script>
</body>
</html>

```

**Comment by patrickhlauke** [comment]
> However, when a button is disabled while the tooltip is being displayed, the tooltip persists.

that's a very edgy edge case that clearly wasn't considered when the code for tooltips was devised, so not surprised it's not working as expected.

**Comment by Johann-S** [comment]
Maybe this section of our docs can help you: https://getbootstrap.com/docs/4.3/components/tooltips/#disabled-elements

**Comment by dagoldman** [comment]
Responding to patrickhlauke - Thanks. In my experience, it is not a "very" edgy case. I admit it does not happen every day, is certainly not the default case. But it happens often enough, and so I think much better if bootstrap could take the case into account.

I think the case could be called "button disables itself". Here is why not so unusual:

Suppose you have a list of items, and a set of buttons to manipulate the items ("add some items", "delete highlighted item", "delete all items", "merge some items", "split highlighted item", "close menu", perhaps others). That's not so unusual. 

Now, when the user clicks "delete all items", the list is cleared, so "delete all items" button should be disabled, as nothing to delete now. But tooltip persists. Another example, after user clicks "split highlighted item", item is split, so "split highlighted item" button should be disabled, as nothing to split now. But tooltip persists. Another case arises when only one item on the list, and user clicks "delete highlighted item". Same situation where button gets disabled, but tooltip persists.

There seem to be other scenarios where "button disables itself" occurs, because I have seen other mentions of "persistent bootstrap tooltip" on internet groups. So it seems this case does occur, is not just a theoretical possibility. 

The discussion on the internet groups centered around writing custom jquery (which I don't really like the syntax of, so have avoided) code using tooltip() function (I take it from popper.js) as a workaround to deal with the problem. That kind of custom programmer by the bootstrap user seems rather complicated and awkward, but it seems for the moment my only choice. If using jquery code is the only method available, is there a recommended way, using my test case as a model?

I do think it would be much better if the bootstrap tooltip would function correctly on it's own, behave the same way as with other buttons, so I hope this can be considered a bug. 

I'm sure the bootstrap developers are very busy. But maybe there is some simple way to add this "button disables itself" case. Or maybe very complex to fix the problem, I do not know. I would volunteer to try helping, but I know really nothing about jquery or popper, so I would be the wrong person to modify the bootstrap source code for this case. 

I would be willing to help with testing or other ways I might assist. Again, thanks for the reply.

**Comment by dagoldman** [comment]
Responding to Johann-S - Thanks, I have looked at the docs and like them. But I am suggesting the tooltip should work correctly by default, not required extra programming. The other buttons "automatically" (I know, somebody programmed all the behaviors) lose their tooltip when they get disabled. But in the "button disables itself" case, the tooltip incorrectly persists (and looks terrible). If I have to do some custom coding using the obscure (to me) jquery syntax, I will fiddle around and do by best, and any specific coding suggestions appreciated. However, it seems much better if bootstrap could take "button disables itself" into account. Thanks again.

**Comment by acshef** [comment]
@dagoldman, my "nuke it from orbit" approach for hiding tooltips is usually something like

```javascript
function hideAllTooltips() {
    [...document.querySelectorAll(".tooltip,.popover")].forEach(tooltipElement => {
        const tooltipParent = $(tooltipElement).data("bs.tooltip").element;
        $(tooltipParent).tooltip("hide");
    });
}
```

And I call that when showing or hiding modals, enabling/disabling buttons, etc. I admit it's a bit heavy-handed, and certainly relies on bootrap/jQuery's internal workings which is a bit of a no-no. But it's also less than 10 lines, so it's not the end of the world to maintain.

**Comment by Johann-S** [comment]
@acshef and you can do : `$(tooltipElement).data("bs.tooltip").hide()` to save one line ;)

**Comment by acshef** [comment]
Thanks @Johann-S, I should have realized that the data stored in `bs.tooltip` was an instance of the Bootstrap class `Tooltip`!

**Comment by dagoldman** [comment]
I appreciate the comments from acshef although I can't make much sense of them or find them very helpful. I confess that may partly be due to my ignorance concerning jquery. Again, I do appreciate the attempt to help.

In any case, I posted my question / bug report separately to stackoverflow and got two helpful answers, one of which seems to work correctly on cross-browser testing. I did multiple tests with with four recent desktop browsers, with iphone safari browser, and with android chrome, and have concluded that the following modification of the change_button function (from the original post) gets rid of the persistent tooltip, at least with my tests:

```
function change_button(button_id, bool_val) {
  var button_obj = document.getElementById(button_id);
  button_obj.disabled = bool_val;
  if (bool_val) {
    $('#' + button_id).tooltip('hide').tooltip('disable');
    }
  else {
    $('#' + button_id).tooltip('enable');
    }
  }
```

I would hope that bootstrap programmers might take the above logic and somehow incorporate into bootstrap so that the tooltip would work correctly by default, not required extra programming. Other buttons lose their tooltip when they get disabled. But in the "button disables itself" case, the tooltip incorrectly persists. To me, that is a bug.

In the meantime, I think I have a method where I can make the tooltips behave correctly. And maybe this post will help others running into the same problem.

If anyone can find a fault with the improved function, or has any other suggestions, please post. Thanks

**Comment by acshef** [comment]
@dagoldman - We're starting to get off topic, but you can optimize your function by passing your `button_obj` to the jQuery `$()` function instead of the unnecessary DOM query*, like so:

```javascript
function change_button(button_id, bool_val) {
  var button_obj = document.getElementById(button_id);
  button_obj.disabled = bool_val;
  if (bool_val) {
    $(button_obj).tooltip('hide').tooltip('disable');
  }
  else {
    $(button_obj).tooltip('enable');
  }
}
```

Regarding my function:
As its name suggests, it's simply a function that hides ALL the currently visible tooltips (and [popovers](https://getbootstrap.com/docs/4.3/components/popovers/)). I don't worry about enabling or disabling the actual functionality of the tooltips, because [once the button is disabled, its tooltip can't easily be shown anyway](https://getbootstrap.com/docs/4.3/components/tooltips/#disabled-elements). I call this function immediately after I disable tooltipable buttons, as well as when I show elements that cover the button (like a popup) that prevent the tooltip from hiding "naturally."

I agree Bootstrap's current behavior could be considered somewhere between "inconsistent" and "incomplete". My solution (like yours) does not modify Bootstrap's behaviors, but rather "patches things up" after the fact.

Footnote:
\* Just like `document.getElementById` requires the engine to search the DOM for the correct element, so too does `$("string query")`. You can get jQuery to skip the (relatively expensive) query by simply passing in the element directly, since you already have a reference to it in `button_obj`.

**Comment by dagoldman** [comment]
acshef - Thank you for improving the change_button function / making it better, and for explaining what your "nuke it from orbit" function does. I see now, just hide all the tooltips, no need to worry which one. Again, I am ignorant concerning jquery. I could learn it if I chose. But I dislike the over terse (to me) / obscure (to me) syntax. Maybe I can convert the change_button function to just use javascript. Or not. Anyway, thanks again.

**Comment by Johann-S** [comment]
That's something we can implement in v5 because `MutationObserver` are supported in IE 11, so I'm in favor to add that feature in v5.

/CC @twbs/js-review

**Comment by alecpl** [comment]
This is a duplicate of #22380.

**Comment by XhmikosR** [comment]
Closing as duplicate of #22380, thanks @alecpl
